### PR TITLE
feat: Add support for splunk HEC endpoints

### PR DESCRIPTION
**Pull request by agrawalx**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)

Part of #8216.

## What's changed and what's your intention?

Adds a **Splunk HTTP Event Collector (HEC) compatible ingestion endpoint**, so existing
Splunk shippers (Vector's `splunk_hec` sink, the OTel Collector `splunk_hec` exporter,
Fluent Bit, etc.) can send logs to GreptimeDB by just changing the URL and token.

**Endpoints** (base path `/v1/splunk`, same convention as `/v1/loki` and `/v1/influxdb`;
the client appends the collector path):

- `POST /v1/splunk/services/collector/event`  events (aliases: `/services/collector`, `.../event/1.0`)
- `GET  /v1/splunk/services/collector/health` health probe (alias: `.../health/1.0`)

gzip request bodies are handled via `RequestDecompressionLayer` (same as Loki).

**How it works**

- **Parsing.** A HEC body is one or more JSON event objects, either concatenated with
  any/no separator or wrapped in a top-level array. A streaming deserializer handles both
  and flattens to a single list of events.
- **Field mapping** per event: `time` -> timestamp, `host`/`source`/`sourcetype` and the
  flat `fields` keys -> tags, `event` (+ any other keys) -> data, `index` ->  table name.
- **Pipeline.** Each event runs through `run_pipeline` with the `greptime_identity`
  default (which flattens a nested `event` object). Overridable per request via the
  `pipeline_name` query param or `x-greptime-pipeline-name` header; a custom pipeline owns
  its own schema.
- **Dynamic tagging.** Identity marks every column `Field`, so we override the metadata
  columns to `Tag` before insert. The insert auto-alter path then adds them to the primary
  key — including brand-new `fields` keys on later writes.
- **Timestamps.** HEC `time` is epoch seconds (optionally fractional); values past ~`1e12`
  are treated as milliseconds; missing/invalid falls back to ingest time.
- **Index -> table name.** Sanitized to a valid table name against the existing
  `NAME_PATTERN` (keeps `-`/`.`/`:`, replaces invalid chars with `_`, fixes the first char);
  empty falls back to `?table=`, then the default `splunk_logs`.
- **Auth.** HEC sends `Authorization: Splunk <token>`; we map the token as `user:pass`
  (split on `:`), reusing the existing InfluxDB `Token`-scheme approach (`Basic` also
  accepted). 
- **HEC response codes:** `0` success, `2` token required, `4` invalid token, `5` no data,
  `6` invalid data, `7` incorrect index, `8` internal; HTTP status mapped from the error.

**Testing**

- Unit tests for the parser (both batch forms), timestamp formats, index sanitization,
  field/tag mapping, and the `Splunk` auth scheme.
- A parser test built from **real captured payloads** of two clients - Vector's
  `splunk_hec` sink and the OTel Collector `splunk_hec` exporter (run in Docker against a
  recording server). They differ on separator, `event` type, and target path, and both
  parse correctly validating the format against real wire output.
- An integration test: real instance + auth -> POST a HEC batch -> query back and assert
  values, assert `host`/`source`/`sourcetype`/`fields` are tags, assert a new
  `fields` key on a later write becomes a tag (dynamic), and assert an invalid `?table=`
  returns code `7`.

**Limitations / out of scope (follow-ups)**

- `/services/collector/raw` (raw-text events).
- Indexer acknowledgment (`X-Splunk-Request-Channel`, `/services/collector/ack`).
- Opaque-token -> identity mapping. Real clients default to a single opaque token (no
  colon), so for now operators must format the token as `user:pass`, or auth fails.

**Compatibility:** purely additive, a new endpoint and new auto-created tables; no
changes to existing APIs, schemas, or data.

## PR Checklist
Please convert it to a draft if some of the following conditions are not met.

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.
- [x] This PR requires documentation updates.
- [x] API changes are backward compatible.
- [x] Schema or data changes are backward compatible.

### CODE REVIEW

File: src/servers/src/http/splunk.rs [hunk 1]
```diff
@@ -0,0 +1,691 @@
...
+        _ => return None,
+    };
+    const MILLIS_THRESHOLD: f64 = 1e12;
+    let nanos = if n >= MILLIS_THRESHOLD {
+        (n * 1000000.0) as i64
+    } else {
+        (n * 1000000000.0) as i64
+    };
+    Some(DateTime::from_timestamp_nanos(nanos))
+}
```
[the remark this answers is not in the document]
Comment (agrawalx) [reply src/servers/src/http/splunk.rs:97]: Resolved this in next commit. the suggestion is applied directly.

File: src/servers/src/http/splunk.rs [hunk 2]
```diff
@@ -0,0 +1,691 @@
...
+pub(crate) const SPLUNK_API_PATH_NAME: &str = "/v1/splunk";
+/// HEC response code for a healthy collector. Splunk returns
+/// `{"text":"HEC is healthy","code":17}`.
+const HEC_HEALTHY_CODE: u32 = 17;
+
+/// HEC response body `{"text", "code"}`; clients branch on `code`.
+/// `use<>`: the response owns `text`, so capture no lifetimes.
+fn hec_response(status: StatusCode, code: u32, text: &str) -> impl IntoResponse + use<> {
+    (status, axum::Json(json!({ "text": text, "code": code })))
+}
```
[the remark this answers is not in the document]
Comment (agrawalx) [reply src/servers/src/http/splunk.rs]: applied.

File: src/servers/src/http/splunk.rs [hunk 3]
```diff
@@ -0,0 +1,691 @@
...
+            continue;
+        };
+        for col in &mut rows.schema {
+            if tag_columns.contains(&col.column_name) {
+                col.semantic_type = SemanticType::Tag as i32;
+            }
+        }
+    }
+    ContextReq::default_opt_with_reqs(reqs)
+}
```
[the remark this answers is not in the document]
Comment (agrawalx) [reply src/servers/src/http/splunk.rs]: applied the fix and added a regression test to exercise exactly this edge case.

File: src/servers/src/http/splunk.rs [hunk 4]
```diff
@@ -0,0 +1,691 @@
...
+    if !first_ok {
+        out.insert(0, '_');
+    }
+
+    NAME_PATTERN_REG.is_match(&out).then_some(out)
+}
+
+pub(crate) fn is_splunk_request<B>(req: &axum::extract::Request<B>) -> bool {
+    req.uri().path().contains(SPLUNK_API_PATH_NAME)
+}
```
[the remark this answers is not in the document]
Comment (agrawalx) [reply src/servers/src/http/splunk.rs:253]: Not sure if this is needed

File: src/servers/src/http/splunk.rs [hunk 5]
```diff
@@ -0,0 +1,691 @@
...
+/// Like `ingest_logs_inner`, but retags metadata columns (identity default) before insert.
+async fn ingest_events(
+    handler: PipelineHandlerRef,
+    pipeline: PipelineDefinition,
+    requests: Vec<PipelineIngestRequest>,
+    query_ctx: QueryContextRef,
+    pipeline_params: GreptimePipelineParams,
+    tag_columns: HashSet<String>,
+    apply_tags: bool,
+) -> Result<HttpResponse> {
```
[the remark this answers is not in the document]
Comment (agrawalx) [reply src/servers/src/http/splunk.rs:263]: applied

File: src/servers/src/http/splunk.rs [hunk 6]
```diff
@@ -0,0 +1,691 @@
...
+    let query_table = params.table.as_deref();
+    let mut by_table: HashMap<String, Vec<VrlValue>> = HashMap::new();
+    let mut tag_columns: HashSet<String> = HashSet::new();
+    for event in events {
+        if let Some((table, map, tags)) = hec_event_to_map(event, DEFAULT_SPLUNK_TABLE, query_table)
+        {
+            tag_columns.extend(tags);
+            by_table.entry(table).or_default().push(map);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (agrawalx) [reply src/servers/src/http/splunk.rs:332]: done.

File: src/servers/src/http/splunk.rs [hunk 7]
```diff
@@ -0,0 +1,753 @@
...
+        DateTime::from_timestamp(secs, nsecs)
+    }
+}
+
+/// Maps one HEC event to `(table, per-event map, tag names)`: `time`->timestamp,
+/// `index`->table, host/source/sourcetype/`fields`->tags, `event`+rest->data.
+/// `None` if the event isn't a JSON object.
+fn hec_event_to_map(
+    event: VrlValue,
+    default_table: &str,
```
Comment (sunng87) [inline src/servers/src/http/splunk.rs]: Is there a chance we will customize this `default_table`? We can inline `DEFAULT_SPLUNK_TABLE` directly

**Comment by killme2008** [comment]
@agrawalx Thanks. Just got back to work, we'll take a look.

File: src/servers/src/http/splunk.rs [hunk 8]
```diff
@@ -0,0 +1,753 @@
...
+        }
+    }
+
+    // host / source / sourcetype are tags.
+    for key in ["host", "source", "sourcetype"] {
+        if let Some(v) = obj.remove(key) {
+            tag_names.push(key.to_string());
+            map.insert(KeyString::from(key), v);
+        }
+    }
```
Comment (sunng87) [inline src/servers/src/http/splunk.rs:183]: I will suggest to put `host`, `source` and `sourcetype` at the beginning of all tags, followed by those customized `fields`

File: src/servers/src/http/splunk.rs [hunk 9]
```diff
@@ -0,0 +1,753 @@
...
+            tag_columns.entry(table.clone()).or_default().extend(tags);
+            by_table.entry(table).or_default().push(map);
+        }
+    }
+    let requests: Vec<PipelineIngestRequest> = by_table
+        .into_iter()
+        .map(|(table, values)| PipelineIngestRequest { table, values })
+        .collect();
+
+    // Bad table name (e.g. invalid `?table=`) -> HEC code 7 ("incorrect index").
```
[the remark this answers is not in the document]
Comment (agrawalx) [reply src/servers/src/http/splunk.rs:343]: added this in next commit.

File: src/servers/src/http/splunk.rs [hunk 10]
```diff
@@ -0,0 +1,753 @@
...
+#[axum_macros::debug_handler]
+pub async fn handle_event(
+    State(log_state): State<LogState>,
+    Query(params): Query<LogIngesterQueryParams>,
+    Extension(query_ctx): Extension<QueryContext>,
+    headers: HeaderMap,
+    payload: Bytes,
+) -> impl IntoResponse {
+    let events = match parse_hec_events(&payload) {
+        Ok(events) => events,
```
[the remark this answers is not in the document]
Comment (agrawalx) [reply src/servers/src/http/splunk.rs:309]: added a channel for Splunk

File: src/servers/src/http/splunk.rs [hunk 11]
```diff
@@ -0,0 +1,753 @@
...
+    if !first_ok {
+        out.insert(0, '_');
+    }
+
+    NAME_PATTERN_REG.is_match(&out).then_some(out)
+}
+
+pub(crate) fn is_splunk_request<B>(req: &axum::extract::Request<B>) -> bool {
+    req.uri().path().contains(SPLUNK_API_PATH_NAME)
+}
```
[the remark this answers is not in the document]
Comment (agrawalx) [reply src/servers/src/http/splunk.rs:253]: changed now. this is still not totally fixed as something like /v1/splunkijk will still pass.

File: src/servers/src/http/splunk.rs (hunk 7 above)
Comment (agrawalx) [reply src/servers/src/http/splunk.rs]: yep, addressed this in latest commit. its inline now and removed the default_table argument.

File: src/servers/src/http/splunk.rs (hunk 8 above)
Comment (agrawalx) [reply src/servers/src/http/splunk.rs:183]: Is this only for better readability? the `ingest_events` calls `run_pipeline` which returns column name in sorted order as `values_to_row` iterates over ObjectMap(a BTreeMap) . 

If host/source/SourceType has to be added in beginning then it would require a change after calling run_pipeline(). changing order here would get overridden when run_pipeline() is called later

Comment (sunng87) [reply src/servers/src/http/splunk.rs:183]: It's also for performance consideration. The order of tags will decide the physical order of data in parquet files. So we prefer to put most queried tag in the front.

File: src/servers/src/http/splunk.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (sunng87) [reply src/servers/src/http/splunk.rs:253]: Can we match for `/v1/splunk/` ? I guess all splunk endpoints have their own subpath?

File: src/servers/src/http/authorize.rs [hunk 12]
```diff
@@ -86,6 +89,14 @@ pub async fn inner_auth<B>(
             crate::metrics::METRIC_AUTH_FAILURE
                 .with_label_values(&[e.status_code().as_ref()])
                 .inc();
+            if is_splunk {
```
Comment (killme2008) [inline src/servers/src/http/authorize.rs]: What about moving `is_splunk` into this code block? Only detect it when errors happen.


```suggestion
            if is_splunk_request(&req) {
```

File: src/servers/src/http/authorize.rs [hunk 13]
```diff
@@ -110,6 +121,10 @@ pub async fn inner_auth<B>(
             crate::metrics::METRIC_AUTH_FAILURE
                 .with_label_values(&[e.status_code().as_ref()])
                 .inc();
+            // HEC: bad credentials -> 4 ("invalid token", 403).
+            if is_splunk {
```
Comment (killme2008) [inline src/servers/src/http/authorize.rs]: The same suggestion as above.

File: src/servers/src/http/splunk.rs [hunk 14]
```diff
@@ -0,0 +1,758 @@
...
+    transform_ndjson_array_factory(values, false)
+}
+
+/// HEC `time`: epoch seconds (optionally fractional); values past ~1e12 are read as
+/// milliseconds. `None` if absent/unparseable (caller falls back to ingest time).
+fn parse_hec_time(value: &VrlValue) -> Option<DateTime<Utc>> {
+    let n: f64 = match value {
+        VrlValue::Integer(i) => *i as f64,
+        VrlValue::Float(f) => f.into_inner(),
+        VrlValue::Bytes(b) => std::str::from_utf8(b).ok()?.trim().parse().ok()?,
```
Comment (killme2008) [inline src/servers/src/http/splunk.rs:81]: Is it possible to be a string?

File: src/servers/src/http/splunk.rs [hunk 15]
```diff
@@ -0,0 +1,758 @@
...
+
+/// Maps one HEC event to `(table, per-event map, tag names)`: `time`->timestamp,
+/// `index`->table, host/source/sourcetype/`fields`->tags, `event`+rest->data.
+/// `None` if the event isn't a JSON object.
+fn hec_event_to_map(
+    event: VrlValue,
+    query_table: Option<&str>,
+) -> Option<(String, VrlValue, Vec<String>)> {
+    let VrlValue::Object(mut obj) = event else {
+        return None;
```
Comment (killme2008) [inline src/servers/src/http/splunk.rs]: Love to add a debug log when it's not a object type.

It's friendly for debugging.

File: src/servers/src/http/splunk.rs [hunk 16]
```diff
@@ -0,0 +1,758 @@
...
+    };
+
+    // Timestamp: HEC `time` is honored first, else ingest time.
+    let ts = obj
+        .remove("time")
+        .as_ref()
+        .and_then(parse_hec_time)
+        .unwrap_or_else(Utc::now);
```
Comment (killme2008) [inline src/servers/src/http/splunk.rs:148]: Should we return an error if time can't be parsed?

Current implementation just use `Utc::now` instead.

File: tests-integration/tests/http.rs [hunk 17]
```diff
@@ -1404,6 +1407,268 @@ pub async fn test_metrics_api(store_type: StorageType) {
...
+        false,
+    )
+    .await;
+    assert_eq!(StatusCode::OK, res.status());
+    assert!(res.text().await.contains("\"code\":0"));
+
+    // 2. Rows landed with the right values (proves route + auth + parse + insert,
+    //    and that the default table name is `splunk_logs`).
+    let rows = get_rows_from_output(
+        &query(
```
Comment (killme2008) [inline tests-integration/tests/http.rs:1530]: Better to select greptime_timestamp too and check its value.

File: src/servers/src/http/splunk.rs [hunk 18]
```diff
@@ -0,0 +1,758 @@
...
+    // host / source / sourcetype are tags.
+    for key in ["host", "source", "sourcetype"] {
+        if let Some(v) = obj.remove(key) {
+            tag_names.push(key.to_string());
+            map.insert(KeyString::from(key), v);
+        }
+    }
+
+    // `event` and any remaining keys are data columns.
+    for (k, v) in obj {
```
Comment (killme2008) [inline src/servers/src/http/splunk.rs:186]: Should we verify that `event` exists and isn't empty?

Looks like splunk will not accept event missing:

* `code 12「Event field is required」` when event is not present
* `code 13「Event field cannot be blank」` when event is empty

**Review by killme2008 — COMMENTED** [review]
Almostly LGTM

Thanks a lot.

File: src/servers/src/http/splunk.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (agrawalx) [reply src/servers/src/http/splunk.rs:253]: Thank you for this suggestion. I will apply

File: src/servers/src/http/authorize.rs (hunk 12 above)
Comment (agrawalx) [reply src/servers/src/http/authorize.rs]: sure, I will apply the suggestion

File: src/servers/src/http/splunk.rs (hunk 15 above)
Comment (agrawalx) [reply src/servers/src/http/splunk.rs]: Thanks, I will add a debug log.

File: src/servers/src/http/splunk.rs (hunk 18 above)
Comment (agrawalx) [reply src/servers/src/http/splunk.rs:186]: oh, I will add this. I missed these error codes.

File: src/servers/src/http/splunk.rs (hunk 16 above)
Comment (agrawalx) [reply src/servers/src/http/splunk.rs:148]: sure, I will change this to a error when time can't be parsed

File: src/servers/src/http/splunk.rs (hunk 14 above)
Comment (agrawalx) [reply src/servers/src/http/splunk.rs:81]: actually the String case is already handled in the `Bytes` arm. This is checked in `parse_time_string_number` test.

File: tests-integration/tests/http.rs (hunk 17 above)
Comment (agrawalx) [reply tests-integration/tests/http.rs:1530]: I added assert for greptime_timestamp and turns out there was a bug. the pipeline overrides the time if a `custom_time_index` is not set and all timestamps were getting overriden to Utc::now. Fixed it now by adding a custom_time_index instead of passing none.

File: src/servers/src/http/splunk.rs (hunk 8 above)
Comment (agrawalx) [reply src/servers/src/http/splunk.rs:183]: Hi, I have made the change for this but I am not sure if this was the preferred approach. Can you please take a look at my last commit, the added test for this reordering passes but I Have added a function inside operator crate in `insert.rs` which only runs for `splunk` requests. I think this is a better approach than making changes in my `apply_tags` function as then the reordering would run for every batch and it will be O(rows x column) operation. 

my approach -> run the reorder once at the time of table creation. all the subsequent requests which come will be reordered correctly by the mito engine.

**Comment by killme2008** [comment]
@agrawalx There are some typos

https://github.com/GreptimeTeam/greptimedb/actions/runs/28205630897/job/83573295728?pr=8321

File: src/servers/src/http/splunk.rs (hunk 8 above)
Comment (killme2008) [reply src/servers/src/http/splunk.rs:183]: @sunng87 Please take a look. I think it's ready to be merged.

**Review by killme2008 — APPROVED** [review]

**Comment by killme2008** [comment]
@agrawalx 
Thanks for your contribution!
